### DISCUSSION TITLE
Use token as password of a PostgreSQL connection and modify it when the token expires, can this be done by a MFlow plugin?

### DISCUSSION THREAD

**Discussion by Ferdinanddb** [mlflow/mlflow#11971] [General]
I am running MLflow on an AKS cluster and I did the necessary so that it is connected to an Azure PostgreSQL Flexible server (to store the metadata) and an Azure storage account (to store the artifacts).

It is working fine, but my ultimate goal is to connect to the PostgreSQL server using an UAMI (Azure managed identity). The advantage of doing so is that, instead of using a password, we ask for a token that expires one hour after, and we perform the operations we want using the token as the password: no need to rotate the password anymore.

I did something similar with Airflow using [an `AirflowPlugin`](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/plugins.html) and [the `sqlalchemy` decorator `event.listens_for(..., "do_connect")`](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DialectEvents.do_connect).

My goal would be to do the same with MLflow, to resume: have some code that would replace the password of the connection with the token value and is executed each time there is an attempt to connect to the PostgreSQL server.

I am not familiar with the MLflow plugins, and I have difficulties to understand their purposes, but are they intended to answer this kind of problem?

If yes, I have the feeling that I should create a class as a child of [this `SqlAlchemyStore` class](https://github.com/mlflow/mlflow/blob/199691a16776a1a6d1e9082c7879bba47fe0e3da/mlflow/store/model_registry/sqlalchemy_store.py#L61) and add my function with my decorator so that the password gets modified whenever the function runs (i.e. each time there is a "do_connect" operation).

If not, what would be the other way to leverage token authentication instead of password authentication?

Thanks if you can help!

**Comment by magnussommarsjo** [comment]
@Ferdinanddb Im in the same situation where I would like to use an Azure managed identity.   
Did you get any conclusion to this?

**Reply by Ferdinanddb** [reply]
Hi,

It has been a while since I did not touch this, but as far as I remember, I found a way to play with this feature by using some specific gunicorn options, something like `--gunicorn-opts --config=/app/gunicorn_config.py`. I defined specific functions inside the `gunicorn_config.py` file named `on_starting`, `post_fork`, `pre_request`, `post_request` to retrieve the token and change the initial value by the retrieved value. All the functions were having similar bodies, like the following:

```py
def pre_request(worker, req):
    import logging
    from sqlalchemy import event, create_engine, Engine
    import logging

    # Set up logging for debugging
    logging.basicConfig(level=logging.DEBUG)

    def inject_azure_token():
        @event.listens_for(Engine, 'do_connect')
        def provide_token(dialect, conn_rec, cargs, cparams):
            logging.debug("Entering provide_token")
            # DO THE NECESSARY TO RETRIEVE THE TOKEN FROM AZURE
            password = ...
            logging.debug(f"Original cparams: {cparams}")
            cparams['password'] = password
            logging.debug(f"Modified cparams: {cparams}")

    def initialize():
        inject_azure_token()
    
    initialize()
    logging.debug(f"Worker {worker.pid} is handling request: {req.uri}")
```

Maybe this function can be improved in lots of way, but I hope it can help + give you inspiration. At the time, I remember I was quite excited by the flexibility that one can get from having such functions that are being automatically executed by gunicorn workers. But to be honest, I was hoping that something more simple was also available than having to deal with gunicorn directly :). 

I had to create my own custom container image, with the `gunicorn_config.py` file and the `mlflow` command to start the server.

I don't know if there is a better way to deal with that now, since I did my tests a year ago, and I don't know if what I did is really stable, although I think it was not bad since what I did is basically forcing the gunicorn worker to execute a function before executing a request (via the function `pre_request`).

Let me know if you don't find a better way, I can send you my code (not really clean though) or try to post it on GitHub.

**Reply by magnussommarsjo** [reply]
Thanks for all the information :) . I will definitely have a look.
